lib/livebook/web_socket/client.ex

```diff
@@ -0,0 +1,220 @@
+        result
+    end
+  end
+
+  defp handle_responses(conn, ref, websocket, [{:data, ref, data}]) do
+    with {:ok, websocket, frames} <- Mint.WebSocket.decode(websocket, data) do
+      case handle_frames(%Response{}, frames) do
+        {:ok, %Response{body: %{type: {:error, _}}} = response} ->
+          {:error, conn, websocket, response}
+
+        {:ok, response} ->
+          {:ok, conn, websocket, response}
+
+        {:close, result} ->
+          disconnect(conn, websocket, ref)
+          {:ok, conn, websocket, result}
+
+        {:error, response} ->
+          {:error, conn, websocket, response}
+      end
+    end
+  end
+
+  defp handle_done_response(conn, ref, response) do
+    case Mint.WebSocket.new(conn, ref, response.status, response.headers) do
+      {:ok, conn, websocket} ->
+        case decode_body(websocket, response) do
+          {websocket, {:ok, result}} ->
+            {:ok, conn, websocket, result}
+
+          {websocket, {:close, result}} ->
+            disconnect(conn, websocket, ref)
+            {:ok, conn, websocket, result}
+
+          {websocket, {:error, reason}} ->
+            {:error, conn, websocket, reason}
+        end
+
+      {:error, conn, %UpgradeFailureError{status_code: status, headers: headers}} ->
+        response = %{
+          response
+          | body: decode_binary(response.body),
+            status: status,
+            headers: headers
+        }
+
+        {:error, conn, response}
+    end
+  end
+
+  defp decode_body(websocket, %Response{status: 101, body: nil}) do
+    {websocket, {:ok, :connected}}
+  end
+
+  defp decode_body(websocket, response) do
+    case Mint.WebSocket.decode(websocket, response.body) do
+      {:ok, websocket, frames} ->
+        {websocket, handle_frames(response, frames)}
+
+      {:error, websocket, reason} ->
+        {websocket, {:error, reason}}
+    end
+  end
+
+  defp decode_binary(binary) when is_binary(binary) do
+    LivebookProto.Response.decode(binary)
+  end
+
+  defp handle_frames(response, frames) do
+    Enum.reduce(frames, response, fn
+      {:binary, binary}, acc ->
+        case decode_binary(binary) do
+          %{type: {:error, _}} = body -> {:error, %{acc | body: body}}
+          body -> {:ok, %{acc | body: body}}
+        end
+
+      {:close, _code, _data}, acc ->
+        {:close, acc}
+    end)
+  end
+
+  @dialyzer {:nowarn_function, send: 4}
+
+  @doc """
+  Sends a message to the given HTTP Connection and WebSocket connection.
+  """
+  @spec send(conn(), websocket(), ref(), frame()) :: send_result()
+  def send(conn, websocket, ref, frame) do
+    with {:ok, websocket, data} <- Mint.WebSocket.encode(websocket, prepare_frame(frame)),
+         {:ok, conn} <- Mint.WebSocket.stream_request_body(conn, ref, data) do
+      {:ok, conn, websocket}
+    end
+  end
+
+  defp prepare_frame(:close), do: :close
+
+  defp prepare_frame(frame) when is_struct(frame) do
+    frame
+    |> LivebookProto.Response.encode()
```

Comment:
Btw, I think I would move the encoding call all the way up to `Livebook.WebSocket.send_request`. This is because once we have a server, we will want to include on the caller, and not inside the Server.